gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -278,40 +319,286 @@ MRCompactorJobPropCreator getJobPropCreator(String topic, Path topicInputDir, Pa
+          if (jobRunner != null) {
+            jobRunner.go();
+          }
+        } else if (getInputFolderState(verifier.inputFolder()) == NO) {
+          LOG.info(String.format("Data incomplete for folder %s. Will try again.", verifier.inputFolder()));
+          if (jobRunner != null) {
+            jobRunner.stop();
+          }
+
+          // Submit another data completeness verification request for this folder with a lower priority.
+          this.execute(
+              getDataComplVerifier(verifier.inputFolder(), verifier.priority() * (1.0 - PRIORITY_REDUCTION_FACTOR)));
+        }
+      } else if (!(r instanceof AbstractDataCompletenessVerifier)) {
+        LOG.error(String.format("Runnable expected to be instance of %s, actual %s",
+            AbstractDataCompletenessVerifier.class.getSimpleName(), r.getClass().getSimpleName()));
+      } else {
+        AbstractDataCompletenessVerifier verifier = (AbstractDataCompletenessVerifier) r;
+        LOG.error("Failed to verify data completeness for " + verifier.inputFolder(), t);
+      }
+    }
+  }
+
+  /**
+   * A subclass of {@link ThreadPoolExecutor} for running compaction jobs, and performs necessary steps
+   * after each compactino job finishes.
+   */
+  private class JobRunnerExecutor extends ThreadPoolExecutor {
+
+    public JobRunnerExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime, TimeUnit unit,
+        BlockingQueue<Runnable> workQueue) {
+      super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
+    }
+
+    /**
+     * When a compaction job finishes, if it successfully published the data (t == null
+     * && jobRunner.status() == COMMITTED), or if it threw throwables (t != null), mark the input folder as complete.
+     * If the job failed to publish the data because the input data was not complete, reduce the priority of
+     * the input folder. A new compaction job will be submitted by the {@link MRJobSubmitter} with a lower priority.
+     */
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+      if (r instanceof MRCompactorJobRunner) {
+        MRCompactorJobRunner jobRunner = (MRCompactorJobRunner) r;
+        MRCompactor.this.jobRunnables.remove(jobRunner.getInputPath().toString());
+        if (t == null) {
+          if (jobRunner.status() == COMMITTED) {
+            LOG.info(String.format("Compacted data for %s successfully published", jobRunner.getInputPath()));
+            setInputFolderState(jobRunner.getInputPath().toString(), COMPLETE);
+          } else {
+            reduceInputFolderPriority(jobRunner.getInputPath().toString());
+          }
+        } else {
+          setInputFolderState(jobRunner.getInputPath().toString(), COMPLETE);
+          MRCompactor.this.compactionJobThrowables.put(jobRunner.getInputPath().toString(), t);
+        }
+      } else {
+        LOG.error(String.format("Runnable expected to be instance of %s, actual %s",
```

Comment:
You probably want to use a `Preconditions.checkArgument()` to make sure the right type is seen here.